triplestore/src/triplestore/__init__.py

```diff
@@ -1,31 +1,148 @@
-# Copyright (C) 2025 Maira Papadopoulou
-# SPDX-License-Identifier: Apache-2.0
-
-
-from typing import Any
-
-from .backends.allegrograph import AllegroGraph
-from .backends.blazegraph import Blazegraph
-from .backends.graphdb import GraphDB
-from .backends.jena import Jena
-from .backends.oxigraph import Oxigraph
-from .base import TriplestoreBackend
-
-_BACKENDS: dict[str, type[TriplestoreBackend]] = {
-    "oxigraph": Oxigraph,
-    "jena": Jena,
-    "graphdb": GraphDB,
-    "blazegraph": Blazegraph,
-    "allegrograph": AllegroGraph,
-}
-
-
-def Triplestore(backend: str, config: dict[str, Any]) -> TriplestoreBackend:
-    """
-    Create an instance of the appropriate triplestore backend.
-    """
-    backend = backend.lower()
-    if backend not in _BACKENDS:
-        msg = f"Unsupported triplestore backend: '{backend}'"
-        raise ValueError(msg)
-    return _BACKENDS[backend](config)
+# Copyright (C) 2025 Maira Papadopoulou
+# SPDX-License-Identifier: Apache-2.0
+
+# Dynamic backend discovery for the `triplestore` package.
+# This module discovers those entry points at runtime and exposes:
+#   - available_backends(): returns the names that can actually be imported now
+#   - Triplestore(name, config): a factory returning a backend instance
+
+from importlib import import_module, metadata
+from typing import Any
+
+from triplestore.base import TriplestoreBackend
+from triplestore.exceptions import BackendNotFoundError, BackendNotInstalledError
+
+REGISTRY: dict[str, str] = {}
+DISCOVERED = False
+
+# Optional mapping used to suggest the correct extra in error messages.
+EXTRA_HINT = {
+    "allegrograph": "allegrograph",
+    "blazegraph": "blazegraph",
+    "graphdb": "graphdb",
+    "jena": "jena",
+    "oxigraph": "oxigraph",
+}
+
+
+def discover_backends() -> None:

```

Comment:
Move the function definitions to a another file (e.g. `registration.py`)
It's convenient to have `__init__.py` files almost empty, besides `import` statements, referring to other files and importing what needs to be exported at module level.